File: Python/146_LRU_Cache.py
#key point: make sure every pre and next point right node

class LinkNode:
    def __init__(self, key, value):
        self.key = key
        self.val = value
        self.pre = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.l = capacity
        self.m = {}
        self.b = LinkNode(None, None)
        self.e = self.b

    def get(self, key: int) -> int:
        if key not in self.m or not self.m[key]:
            return -1
        if self.m[key].next:
            self.m[key].pre.next = self.m[key].next
            self.m[key].next.pre = self.m[key].pre
            self.e.next = self.m[key]
            self.m[key].pre = self.e
            self.e = self.e.next
            self.e.next = None
        return self.m[key].val

    def put(self, key: int, value: int) -> None:
        if key not in self.m or not self.m[key]:
            if self.l > 0:
                self.e.next = LinkNode(key, value)
                self.e.next.pre = self.e
                self.e = self.e.next
                self.m[key] = self.e
                self.l -= 1
            else:
                self.b = self.b.next
                self.m[self.b.key] = None
                self.e.next = LinkNode(key, value)
                self.e.next.pre = self.e
                self.e = self.e.next
                self.m[key] = self.e
        else:
            if self.m[key].next:
                self.m[key].next.pre = self.m[key].pre
                self.m[key].pre.next = self.m[key].next
                self.e.next = self.m[key]
                self.e.next.pre = self.e
                self.e = self.e.next
                self.e.next = None
            self.m[key].val = value

File: leetcode/archived/golang/146_LRU_Cache.go
package main

type linknode struct{
	key, value int
	pre, next *linknode
}

type LRUCache struct {
	capacity int
	m map[int] *linknode
	b, e *linknode
}


func Constructor(capacity int) LRUCache {
	lru := LRUCache{capacity: capacity}
	lru.b = &linknode{key: 0, value:0}
	lru.m = make(map[int] *linknode)
	lru.e = lru.b
	return lru
}


func (this *LRUCache) Get(key int) int {
	if _, ok:= this.m[key]; !ok{
		return -1
	}
	if this.m[key].next != nil{
		this.m[key].next.pre = this.m[key].pre
		this.m[key].pre.next = this.m[key].next
		this.e.next = this.m[key]
		this.m[key].pre = this.e
		this.e = this.e.next
		this.e.next = nil
	}
	return this.m[key].value
}


func (this *LRUCache) Put(key int, value int)  {
	if _, ok := this.m[key]; !ok{
		if this.capacity > 0{
			this.e.next = &linknode{key: key, value:value, pre:this.e, next:nil}
			this.e = this.e.next
			this.m[key] = this.e
			this.capacity--
		}else{
			this.b = this.b.next
			delete(this.m, this.b.key)
			this.e.next = &linknode{key: key, value:value, pre:this.e, next:nil}
			this.e = this.e.next
			this.m[key] = this.e
		}
	}else{
		if this.m[key].next != nil{
			this.m[key].next.pre = this.m[key].pre
			this.m[key].pre.next = this.m[key].next
			this.e.next = this.m[key]
			this.e.next.pre = this.e
			this.e = this.e.next
			this.e.next = nil
		}
		this.m[key].value = value
	}
}


/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */


